Dynamic decision to swap threads or not on Erlang Write operation

matthewvon opened the pull request:
eleveldb swaps each Erlang request from an Erlang scheduler thread to a local eleveldb worker thread.  This is to prevent Erlang scheduler collapses.  Write operations do not actually need to swap all the time.  This branch creates the capability of dynamic selection to swap threads or not.

Detailed discussion is here:

https://github.com/basho/leveldb/wiki/mv-nonblocking-write2

Commit:
mods from macintux to allow write to be both sync and async.

Commit:
temporary change to tie this eleveldb branch to leveldb branch mv-nonblocking-write

Commit:
modify async_write() to have two code paths: one non-blocking path into leveldb without thread swap and second using thread swap when leveldb predicted to block

Commit:
move debug performance counters.

macintux committed:
Fix test invocation of async_put/5

`async_put` now can return:
* context value indicating a `receive` is appropriate to wait for a
  follow-up message
* `ok` if the write went straight through
* error tuple

Commit:
document return characteristics of async_write

paulplace commented on c_src/build_deps.sh:
```diff
@@ -8,7 +8,7 @@ if [ `uname -s` = 'SunOS' -a "${POSIX_SHELL}" != "true" ]; then
 unset POSIX_SHELL # clear it so if we invoke other scripts, they run as ksh as well
 
-LEVELDB_VSN=""
+LEVELDB_VSN="mv-nonblocking-write"
```
Standard reminder to set this back to "" before merging.

paulplace commented on c_src/eleveldb.cc:
```diff
@@ -703,40 +717,67 @@ async_write(
+        eleveldb_priv_data& data = *static_cast<eleveldb_priv_data *>(enif_priv_data(env));
+        if(false == data.thread_pool.Submit(work_item))
+        {
+            // work_item contains "batch" and the delete below gets both memory allocations
```
work_item also takes ownership of the opts variable that was allocated on the heap; it's probably worth updating the comment so that the next person in this code doesn't have to hunt down that fact

macintux committed:
Support legacy use of `async_put/5`

Previous commits broke the previous `async_put` API; this restores the
old behavior while adding a new `async_put2/5` interface for future
releases of Riak.

See commentary for explanation.

macintux committed:
Merge branch 'mv-nonblocking-write2' of github.com:basho/eleveldb into mv-nonblocking-write2

matthewvon commented:
this branch was replaced by mv-nonblocking-writes3 ... a manual rebase
